<!DOCTYPE html>
<html lang="en" class="js csstransforms3d">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no">
    <meta name="generator" content="Hugo 0.17" />
    <meta name="description" content="">


    <link rel="shortcut icon" href="https://epiviz.github.io/tutorials/images/favicon.png" type="image/x-icon" />

    
    <title>Longitudinal Data Analysis</title>
    <link href="https://epiviz.github.io/tutorials/css/nucleus.css" rel="stylesheet">
    <link href="https://epiviz.github.io/tutorials/css/font-awesome.min.css" rel="stylesheet">
    <link href="https://epiviz.github.io/tutorials/css/hybrid.css" rel="stylesheet">
    <link href="https://epiviz.github.io/tutorials/css/featherlight.min.css" rel="stylesheet">
    <link href="https://epiviz.github.io/tutorials/css/perfect-scrollbar.min.css" rel="stylesheet">
    <link href="https://epiviz.github.io/tutorials/css/horsey.css" rel="stylesheet">
    <link href="https://epiviz.github.io/tutorials/css/theme.css" rel="stylesheet">
    <link href="https://epiviz.github.io/tutorials/css/hugo-theme.css" rel="stylesheet">
    <script src="https://epiviz.github.io/tutorials/js/jquery-2.x.min.js"></script>
    <script src="https://epiviz.github.io/tutorials/js/prism.js"></script>
    <style type="text/css">:root #header + #content > #left > #rlblock_left
    {display:none !important;}</style>
    <link href="https://epiviz.github.io/tutorials/css/my_style.css" rel="stylesheet">
  </head>
  <body class="" data-url="/metaviz/spline/">
    <nav id="sidebar">
  <div id="header-wrapper">
    <div id="header">
      <a width="100%" height="100%" href="https://epiviz.github.io"><img src="https://epiviz.github.io/tutorials/images/epiviz_logo.png"></a>
    </div>
    
</div>


  <div class="highlightable">
    <ul class="topics">
      
      
        
      
      
      
      

      
        
      
      
      

      <li class="dd-item  parent" data-nav-id="/metaviz/">
        <a href="https://epiviz.github.io/tutorials/metaviz/">
          <span>
            
              <b>X. </b>
            
             Metaviz Tutorials
            
           </span>
        </a>
        
        <ul>
          
          
          
          
            <li class="dd-item " data-nav-id="/metaviz/overview/">
              <a href="https://epiviz.github.io/tutorials/metaviz/overview/">
                <span>Metaviz Quick Overview     </i></span>
              </a>
            </li>
          
          
          
            <li class="dd-item " data-nav-id="/metaviz/metagenomeSeq/">
              <a href="https://epiviz.github.io/tutorials/metaviz/metagenomeSeq/">
                <span>Differential Analysis with MetagenomeSeq and Metaviz     </i></span>
              </a>
            </li>
          
          
          
            <li class="dd-item " data-nav-id="/metaviz/video_tutorials/">
              <a href="https://epiviz.github.io/tutorials/metaviz/video_tutorials/">
                <span>Visual Analysis of MSD16S Dataset     </i></span>
              </a>
            </li>
          
          
          
            <li class="dd-item active" data-nav-id="/metaviz/spline/">
              <a href="https://epiviz.github.io/tutorials/metaviz/spline/">
                <span>Longitudinal Data Analysis     </i></span>
              </a>
            </li>
          
          
          
            <li class="dd-item " data-nav-id="/metaviz/usingDocker/">
              <a href="https://epiviz.github.io/tutorials/metaviz/usingDocker/">
                <span>Metaviz Docker Containers     </i></span>
              </a>
            </li>
          
          
          
            <li class="dd-item " data-nav-id="/metaviz/hostingData/">
              <a href="https://epiviz.github.io/tutorials/metaviz/hostingData/">
                <span>Hosting Data in Metaviz     </i></span>
              </a>
            </li>
          
          
        </ul>
        
      </li>
      
      
      

      
        
      
      
      

      <li class="dd-item  " data-nav-id="/tutorials/">
        <a href="https://epiviz.github.io/tutorials/tutorials/">
          <span>
            
              <b>X. </b>
            
             Epiviz Tutorials
            
           </span>
        </a>
        
        <ul>
          
          
          
          
            <li class="dd-item " data-nav-id="/tutorials/timp2014/">
              <a href="https://epiviz.github.io/tutorials/tutorials/timp2014/">
                <span>450k Illumina Human Methylation data for multiple solid tumors     </i></span>
              </a>
            </li>
          
          
          
            <li class="dd-item " data-nav-id="/tutorials/charts/">
              <a href="https://epiviz.github.io/tutorials/tutorials/charts/">
                <span>Chart Tutorial     </i></span>
              </a>
            </li>
          
          
          
            <li class="dd-item " data-nav-id="/tutorials/computed-measurements/">
              <a href="https://epiviz.github.io/tutorials/tutorials/computed-measurements/">
                <span>Computed Measurements     </i></span>
              </a>
            </li>
          
          
          
            <li class="dd-item " data-nav-id="/tutorials/epivizr/">
              <a href="https://epiviz.github.io/tutorials/tutorials/epivizr/">
                <span>Epivizr     </i></span>
              </a>
            </li>
          
          
          
            <li class="dd-item " data-nav-id="/tutorials/plugins/">
              <a href="https://epiviz.github.io/tutorials/tutorials/plugins/">
                <span>External Scripts and Plugins     </i></span>
              </a>
            </li>
          
          
          
            <li class="dd-item " data-nav-id="/tutorials/websocket/">
              <a href="https://epiviz.github.io/tutorials/tutorials/websocket/">
                <span>Websocket Tutorial     </i></span>
              </a>
            </li>
          
          
        </ul>
        
      </li>
      
      
    </ul>
    <hr>
     
    <section id="footer">
      <p>Built with <a href="https://github.com/matcornic/hugo-theme-learn"><i class="fa fa-heart"></i></a> from <a href="http://getgrav.org">Grav</a> and <a href="http://gohugo.io/">Hugo</a></p>
    </section>
  </div>
</nav>

        <section id="body">
        <div id="overlay"></div>

        <div class="padding highlightable">

            <div id="top-bar">
              
              <div id="breadcrumbs" itemscope="" itemtype="http://data-vocabulary.org/Breadcrumb">
                  <span id="sidebar-toggle-span">
                      <a href="#" id="sidebar-toggle" data-sidebar-toggle="">
                        <i class="fa fa-bars"></i>
                      </a>
                  </span>
                
                <span id="toc-menu"><a href=""><i class="fa fa-list-alt"></i></a></span>
                
                
                
                
                  
                    
                    
                <a href="https://epiviz.github.io/tutorials/metaviz/" itemprop="url"><span itemprop="title">Metaviz Tutorials</span></a> <i class="fa fa-angle-right"></i>
                    
                  
                
                  
                
                <span itemprop="title"> Longitudinal Data Analysis</span>
              </div>
              
                  <div class="progress">
    <div class="wrapper">
<nav id="TableOfContents">
<ul>
<li>
<ul>
<li><a href="#generating-metagenomeseq-objects-and-performing-ss-anova-testing">Generating metagenomeSeq objects and performing SS-ANOVA testing</a></li>
<li><a href="#generating-metavizr-object-to-plot">Generating metavizR object to plot</a></li>
<li><a href="#choosing-features-to-plot-and-adding-a-line-plot-to-metaviz">Choosing features to plot and adding a line plot to Metaviz</a></li>
<li><a href="#updating-chart-settings-from-metavizr">Updating chart settings from metavizR</a></li>
</ul></li>
</ul>
</nav>
    </div>
</div>

              

            </div>
            
    	        <div id="body-inner">
                
                <h1>Longitudinal Data Analysis</h1>
                



<p>Metaviz is a tool for interactive visualization and exploration of metagenomic sequencing data. It provides a novel navigation tool for exploring hierarchical feature data that is coupled with multiple data visualizations including heatmaps, stacked bar charts, and scatter plots.
The metavizr package implements two-way communication between the R/Bioconductor computational genomics environment and Metaviz. Objects in an R/Bioconductor session can be visualized and explored using the Metaviz navigation tool and plots.
In this post we show how to use metavizr and the metagenomeSeq Bioconductor package to
analyze data from a longitudinal study.</p>

<h2 id="generating-metagenomeseq-objects-and-performing-ss-anova-testing">Generating metagenomeSeq objects and performing SS-ANOVA testing</h2>

<p>We use a dataset from &ldquo;Individual-specific changes in the human gut microbiota after challenge with enterotoxigenic Escherichia coli and subsequent ciprofloxacin treatment&rdquo; which includes data from 12 human samples that were challenged with E. coli then provided subsequent antibiotic treatment.<br />
Samples were taken for 1 day pre-infection and 10 days post-infection. We will analyze this time series data with a smoothing-spline and plot it in a Metaviz workspace with a line plot over all features.</p>

<p>First, import the etec16s dataset, select sample data from the first 9 days. Then use metavizr to launch an instance of the Metaviz app in a local web-browser.</p>

<pre><code class="language-r, eval=FALSE">library(etec16s)
library(metagenomeSeq)
library(metavizr)
data(etec16s)
etec16s &lt;- etec16s[,-which(pData(etec16s)$Day&gt;9)]

app &lt;- startMetaviz(&quot;http://metaviz.cbcb.umd.edu&quot;) 
</code></pre>

<p>This provides a new workspace.</p>

<p><img src="https://epiviz.github.io/tutorials/images/metaviz/SplineAppLaunch.png" alt="" /></p>

<p>Next, use metagenomeSeq to fit a smoothing-spline to the time series data using the &lsquo;timeSeriesFits&rsquo; method.</p>

<pre><code class="language-r, eval=FALSE">featureData(etec16s)$Kingdom &lt;- &quot;Bacteria&quot;
feature_order &lt;- c(&quot;Kingdom&quot;, &quot;Phylum&quot;, &quot;Class&quot;, &quot;Order&quot;, &quot;Family&quot;, &quot;Genus&quot;, &quot;Species&quot;, &quot;OTU.ID&quot;)

timeSeriesFits &lt;- fitMultipleTimeSeries(obj=etec16s,
                             formula = abundance~id + time*class + AntiGiven,
                             class=&quot;AnyDayDiarrhea&quot;,
                             id=&quot;SubjectID&quot;,
                             time=&quot;Day&quot;,
                             lvl=&quot;Species&quot;,
                             feature_order = feature_order,
                             C=0.3,
                             B=1)
</code></pre>

<h2 id="generating-metavizr-object-to-plot">Generating metavizR object to plot</h2>

<p>For plotting the data using Metaviz, we use the fit values as y-coordinates and timepoints as x-coordinates.  We need to call &lsquo;ts2MRexperiment&rsquo; with arguments for the sample and feature data, in this case timepoints and annotations, respectively. We also
aggregate to the &lsquo;species&rsquo; level.</p>

<pre><code class="language-r, eval=FALSE">feature_order2 &lt;- c(&quot;Kingdom&quot;, &quot;Phylum&quot;, &quot;Class&quot;, &quot;Order&quot;, &quot;Family&quot;, &quot;Genus&quot;, &quot;Species&quot;)

splinesMRexp &lt;- ts2MRexperiment(timeSeriesFits, feature_data = featureData(aggregateByTaxonomy(etec16s, lvl=&quot;Species&quot;, feature_order = feature_order2)), sampleNames = timeSeriesFits[[2]]$fit$timePoints)
etec16s_species &lt;- aggregateByTaxonomy(etec16s, lvl=&quot;Species&quot;, feature_order = feature_order)
</code></pre>

<h2 id="choosing-features-to-plot-and-adding-a-line-plot-to-metaviz">Choosing features to plot and adding a line plot to Metaviz</h2>

<p>We select features with a timepoint with absolute log-fold change greater than 2. We add the MRexperiment as a measurement for Metaviz to plot the hierarchy in a &lsquo;FacetZoom&rsquo; object.</p>

<pre><code class="language-r, eval=FALSE">splines_to_plot &lt;- sapply(1:nrow(MRcounts(splinesMRexp)), function(i) {max(abs(MRcounts(splinesMRexp[i,]))) &gt;= 2})
splines_to_plot_indices &lt;- which(splines_to_plot == TRUE)

ic_plot &lt;- app$plot(splinesMRexp[splines_to_plot_indices,], datasource_name = &quot;etec16_base&quot;, control=metavizControl(norm = FALSE, aggregateAtDepth = 6), feature_order = feature_order2)
</code></pre>

<p><img src="https://epiviz.github.io/tutorials/images/metaviz/SplineAddIcicle.png" alt="" /></p>

<p>We then use the &lsquo;revisualize&rsquo; method to add a Line Plot for the smoothing spline.</p>

<pre><code class="language-r, eval=FALSE">splineObj &lt;- app$data_mgr$add_measurements(splinesMRexp[splines_to_plot_indices,], datasource_name = &quot;etec16_base&quot;, control = metavizControl(norm=FALSE, aggregateAtDepth = 6))
splineMeasurements &lt;- splineObj$get_measurements()
splineChart &lt;- app$chart_mgr$visualize(&quot;LinePlot&quot;, splineMeasurements)
</code></pre>

<p>This adds the Line Plot to the workspace but we will need to set the range of the y-axis and modify the color scheme to produce a final workspace.</p>

<p><img src="https://epiviz.github.io/tutorials/images/metaviz/SplineLinePlotAdded.png" alt="" /></p>

<h2 id="updating-chart-settings-from-metavizr">Updating chart settings from metavizR</h2>

<p>We can update the colors and settings on the spline chart. For example, lets limit the y axis to be between -10 and 10. To do so we use the &lsquo;set_chart_settings&rsquo; method. We can list existing settings for a chart using the &lsquo;list_chart_settings&rsquo; function.</p>

<pre><code class="language-r, eval=FALSE, echo=FALSE"># list available charts
app$chart_mgr$list_chart_types()

# list available settings for &quot;LinePlot&quot;
app$chart_mgr$list_chart_type_settings(&quot;LinePlot&quot;)

# update settings on splineChart 
settings &lt;- list(yMin = -10, yMax = 10)
colors &lt;- c(&quot;#1f77b4&quot;, &quot;#ff7f0e&quot;, &quot;#2ca02c&quot;, &quot;#d62728&quot;, &quot;#9467bd&quot;, &quot;#8c564b&quot;, &quot;#e377c2&quot;, &quot;#7f7f7f&quot;, &quot;#bcbd22&quot;, &quot;#17becf&quot;)
app$chart_mgr$set_chart_settings(splineChart, settings=settings, colors = colors)
</code></pre>

<p><img src="https://epiviz.github.io/tutorials/images/metaviz/SplineLinePlotSettings.png" alt="" /></p>

<p>We can interact with the Metaviz session, such as hovering over the Line Plot to see highlight the path through the hierarchy.</p>

<p><img src="https://epiviz.github.io/tutorials/images/metaviz/SplineLinePlotHover.png" alt="" /></p>

<p>Also, we can hover in the icicle plot to highlight all features under that path in the Line Plot.</p>

<p><img src="https://epiviz.github.io/tutorials/images/metaviz/SplineIcicleHover.png" alt="" /></p>


      
      </div>
    </div>

    <div id="navigation">
        
        
        
    </div>

    </section>
    <div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
      <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
    </div>
    <script src="https://epiviz.github.io/tutorials/js/clipboard.min.js"></script>
    <script src="https://epiviz.github.io/tutorials/js/perfect-scrollbar.min.js"></script>
    <script src="https://epiviz.github.io/tutorials/js/perfect-scrollbar.jquery.min.js"></script>
    <script src="https://epiviz.github.io/tutorials/js/jquery.sticky-kit.min.js"></script>
    <script src="https://epiviz.github.io/tutorials/js/featherlight.min.js"></script>
    <script src="https://epiviz.github.io/tutorials/js/html5shiv-printshiv.min.js"></script>
    <script src="https://epiviz.github.io/tutorials/js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="https://epiviz.github.io/tutorials/js/modernizr.custom.71422.js"></script>
    <script src="https://epiviz.github.io/tutorials/js/learn.js"></script>
    <script src="https://epiviz.github.io/tutorials/js/hugo-learn.js"></script>
    

  </body>
</html>

