<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to metavizr: interactive visualization for metagenomics • metavizr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">metavizr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Introduction to metavizr: interactive visualization for metagenomics</h1>
                        <h4 class="author">Héctor Corrada Bravo, Florin Chelaru, Justin Wagner, Jayaram Kancherla, Joseph Paulson</h4>
            
            <h4 class="date">2017-02-02</h4>
          </div>

    
    
<div class="contents">
<p><a href="http://metaviz.cbcb.umd.edu"><code>Metaviz</code></a> is tool for interactive visualization and exploration of metagenomic sequencing data. Metaviz a provides main navigation tool for exploring hierarchical feature data that is coupled with multiple data visualizations including heatmaps, stacked bar charts, and scatter plots. Metaviz supports a flexible plugin framework so users can add new <a href="http://d3js.org/">d3</a> visualizations. You can find more information about Metaviz at <a href="http://metaviz.cbcb.umd.edu/help" class="uri">http://metaviz.cbcb.umd.edu/help</a>.</p>
<p>The <code>metavizr</code> package implements two-way communication between the <code>R/Bioconductor</code> computational genomics environment and <code>Metaviz</code>. Objects in an <code>R/Bioconductor</code> session can be visualized and explored using the <code>Metaviz</code> navigation tool and plots. Metavizr uses Websockets to communicate between the browser Javascript client and the <code>R/Bioconductor</code> session. Websockets are the protocols underlying the popular <a href="http://www.rstudio.com/shiny/">Shiny</a> system for authoring interactive web-based reports in R.</p>
<div id="preliminaries-the-data" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#preliminaries-the-data" class="anchor"> </a></body></html>Preliminaries: the data</h1>
<p>In this vignette we will look at two data sets one from a case/control study and another with a time series. We load the first data set from the <code>msd16s</code> Bioconductor package. This data is from the Moderate to Severe Diaherrial disease study in children from four countries: Banglash, Kenya, Mali, and the Gambia. Case and control stool samples were gathered from each country across several age ranges, 0-6 months, 6-12 months, 12-18 months, 18-24 months, and 24-60 months. An analysis of this data is described in Pop et al. [1].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(metavizr)
<span class="kw">require</span>(metagenomeSeq)
<span class="kw">require</span>(msd16s)</code></pre></div>
</div>
<div id="the-metavizr-session-manager" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#the-metavizr-session-manager" class="anchor"> </a></body></html>The metavizr session manager</h1>
<p>The connection to <code>Metaviz</code> is managed through a session manager object of class <code>EpivizApp</code>. We can create this object and open <code>Metaviz</code> using the <code>startMetaviz</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app &lt;-<span class="st"> </span><span class="kw"><a href="../reference/startMetaviz.html">startMetaviz</a></span>()</code></pre></div>
<p>This opens a websocket connection between the interactive <code>R</code> session and the browser client. This will allow us to visualize data stored in the <code>Metaviz</code> server along with data in the interactive <code>R</code> session.</p>
<hr>
<p><em>Windows users:</em> In Windows platforms we need to use the <code>service</code> function to let the interactive <code>R</code> session connect to the <code>epiviz</code> web app and serve data requests. We then escape (using <code>ctl-c</code> or <code>esc</code> depending on your environment) to continue with the interactive <code>R</code> session. This is required anytime you want <code>metavizr</code> to serve data to the web app, for example, when interacting with the UI. (We are actively developing support for non-blocking sessions in Windows platforms).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">app$server$<span class="kw">service</span>()</code></pre></div>
<hr>
<p>For vignette purposes, we will subset the 992 <code>msd16s</code> samples to those 301 from Bangladesh. Also, we will aggregate the count matrix to the species level. We have found that matrix sizes with YYY elements perform well for interactive visualization with an R session using WebSockets. For larger abundance matrices, we recommend using the graph database backend available at [NEED TO ADD LINK]. Once having subset the data, we normalize count data using the <code>metagenomeSeq::cumNorm</code> function, choose genera that have more than 5 OTUs in the dataset and select the 10 genera with largest standard deviation of log-transformed normalized counts across samples. Set aggregation level to “class”, will find features DA at this level and then change an push those changes to an MRexperiment that is aggreagted to “species” level, and then explore the hiearchy.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">feature_order &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"superkingdom"</span>, <span class="st">"phylum"</span>, <span class="st">"class"</span>, <span class="st">"order"</span>, <span class="st">"family"</span>, <span class="st">"genus"</span>, <span class="st">"species"</span>, <span class="st">"OTU"</span>)
aggregated_feature_order &lt;-<span class="st"> </span>feature_order[<span class="dv">1</span>:<span class="dv">7</span>]

msd16s_species &lt;-<span class="st"> </span>msd16s
<span class="kw">fData</span>(msd16s) &lt;-<span class="st"> </span><span class="kw">fData</span>(msd16s)[feature_order]
<span class="kw">fData</span>(msd16s_species) &lt;-<span class="st"> </span><span class="kw">fData</span>(msd16s_species)[aggregated_feature_order]
  
bangladesh &lt;-<span class="st"> </span>msd16s[, <span class="kw">which</span>(<span class="kw">pData</span>(msd16s)$Country ==<span class="st"> "Bangladesh"</span>)]
bangladesh_species &lt;-<span class="st"> </span>msd16s_species[, <span class="kw">which</span>(<span class="kw">pData</span>(msd16s_species)$Country ==<span class="st"> "Bangladesh"</span>)]

aggregated_species &lt;-<span class="st">  </span><span class="kw">cumNorm</span>(<span class="kw">aggregateByTaxonomy</span>(bangladesh_species, <span class="dt">lvl=</span><span class="st">"species"</span>), <span class="dt">p =</span> <span class="fl">0.75</span>)

aggregation_level &lt;-<span class="st"> "class"</span>
aggregated_bangladesh &lt;-<span class="st"> </span><span class="kw">aggregateByTaxonomy</span>(bangladesh, <span class="dt">lvl=</span>aggregation_level)

normed_bangladesh &lt;-<span class="st">  </span><span class="kw">cumNorm</span>(aggregated_bangladesh, <span class="dt">p =</span> <span class="fl">0.75</span>)
bangladesh_sample_data &lt;-<span class="st">  </span><span class="kw">pData</span>(normed_bangladesh)
mod &lt;-<span class="st">  </span><span class="kw">model.matrix</span>(~<span class="dv">1</span>+Dysentery, <span class="dt">data =</span> bangladesh_sample_data)
results_bangladesh &lt;-<span class="st">  </span><span class="kw">fitFeatureModel</span>(normed_bangladesh, mod)</code></pre></div>
<pre><code>## Warning: glm.fit: algorithm did not converge</code></pre>
<pre><code>## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">logFC_bangladesh &lt;-<span class="st"> </span><span class="kw">MRcoefs</span>(results_bangladesh, <span class="dt">number =</span> <span class="kw">nrow</span>(normed_bangladesh))

feature_names &lt;-<span class="st"> </span><span class="kw">rownames</span>(logFC_bangladesh[<span class="kw">which</span>(logFC_bangladesh[<span class="kw">which</span>(<span class="kw">abs</span>(logFC_bangladesh$logFC) &gt;<span class="st"> </span><span class="dv">2</span>),]$adjPvalues &lt;<span class="st"> </span>.<span class="dv">05</span>),]) 
fSelection &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateSelection.html">generateSelection</a></span>(<span class="dt">feature_names =</span>feature_names , <span class="dt">aggregation_level =</span> aggregation_level, <span class="dt">selection_type =</span><span class="dv">2</span>)

control &lt;-<span class="st"> </span>metavizr::<span class="kw"><a href="../reference/metavizControl.html">metavizControl</a></span>(<span class="dt">featureSelection =</span> fSelection)</code></pre></div>
</div>
<div id="adding-an-icicle-navigation-widget-to-explore-metagenomic-features" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#adding-an-icicle-navigation-widget-to-explore-metagenomic-features" class="anchor"> </a></body></html>Adding an Icicle navigation widget to explore metagenomic features</h1>
<p>Once the browser is open we can visualize the metagenomic features from the <code>mouseData</code> object. We use the <code>plot</code> method to do so.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#specify the feature hierarchy for the dataset</span>
icicle_plot &lt;-<span class="st"> </span>app$<span class="kw">plot</span>(aggregated_species, <span class="dt">datasource_name=</span><span class="st">"mmssdd"</span>, <span class="dt">control=</span>control, <span class="dt">feature_order =</span> aggregated_feature_order)</code></pre></div>
<pre><code>## MRExperiment Object validated... PASS</code></pre>
<p>You should now see an icicle widget to explore the hierarchy of the metagenomic features from the <code>mouseData</code> object. To navigate the complex, hierarchical structure of the feature space, we developed an icicle/facet zoom visualization. Because of the limitations in the screen size and performance rendering big trees, the icicle plot is an efficient visualization for navigating trees. The icicle visualization helps zoom in and out of trees and traverse subtrees. Every node in the icicle has a state associated with it. There are three possible states for a node 1) expand - use all subtree nodes during analysis 2) collapse - aggregate all nodes under the subtree to the selected node 3) remove - discard all nodes in the subtree for the analysis. The state of a node is also propagated to all its children and can be identified by the opacity of the node. Row level controls are available to the left of the icicle - to set the state of all nodes at a selected depth/taxonomic class of the hierarchy. Icicle is traversible if the ends on the row controls are chevrons. Users can set states on the nodes to define a cut over the feature space. The cut defines how the count data is queried, analyzed and visualized in other plots like heat maps or stacked line plots. In addition to defining the cut, we also have a navigation bar on top of the icicle to limit the range of features when querying for count data. Navigation bar is a flexible component (increase/decrease the length) and controls are available to move the bar left/right and extend over the entire range of the current tree in the icicle. Each of these actions would query the count data and automatically propagate the changes to other visualizations in the workspace. When navigating outside the scope of the navigation bar, chevrons (left/right) appear on the navigation bar to help identify the current position. Brushing/hovering is another important visual element we focused on our implementation of an icicle. Hovering over a node hovers its path in the tree i.e. highlight both its parents and children. When there are other visualizations like heat maps or stacked line plots in the workspace, hovering over a chart element highlights the mapped feature nodes in all the other plots.</p>
</div>
<div id="visualizing-count-data-from-the-mousedata-mrexperiment-object" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#visualizing-count-data-from-the-mousedata-mrexperiment-object" class="anchor"> </a></body></html>Visualizing count data from the <code>mouseData</code> MRExperiment object</h1>
<p>Now we can view the data as a heatmap calling <code>revisualize</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">heatmap &lt;-<span class="st"> </span>app$chart_mgr$<span class="kw">revisualize</span>(<span class="dt">chart_type =</span> <span class="st">"HeatmapPlot"</span>, <span class="dt">chart =</span> icicle_plot)</code></pre></div>
<p>Using the same data, we can also revisualize it in a stacked plot to see the abundance of various features across samples. Since the measurements are added from creating the icicle_plot, we only need to add a stacked line plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">stackedPlot &lt;-<span class="st"> </span>app$chart_mgr$<span class="kw">revisualize</span>(<span class="dt">chart_type =</span><span class="st">"StackedLinePlot"</span>, <span class="dt">chart =</span> icicle_plot)</code></pre></div>
<p>Finally, we can update the threshold cutoff we had for fold change, pass those modifications the icicle plot, and see the updates propogate to the heat map and stacked plot. This shows the use case of statistically-guided interactive visualizations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">feature_names2 &lt;-<span class="st"> </span><span class="kw">rownames</span>(logFC_bangladesh[<span class="kw">which</span>(logFC_bangladesh[<span class="kw">which</span>(<span class="kw">abs</span>(logFC_bangladesh$logFC) &gt;<span class="st"> </span>.<span class="dv">5</span>),]$adjPvalues &lt;<span class="st"> </span>.<span class="dv">05</span>),])
fSelection2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateSelection.html">generateSelection</a></span>(<span class="dt">feature_names =</span>  feature_names2, <span class="dt">aggregation_level =</span> aggregation_level, <span class="dt">selection_type =</span><span class="dv">2</span>)

app$<span class="kw">get_ms_object</span>(icicle_plot)$<span class="kw">featureSelection</span>(<span class="dt">featureNames=</span>fSelection2$featureNames, <span class="dt">featureOrder=</span>fSelection2$featureOrder, <span class="dt">featureLevel=</span>fSelection2$featureLevel, <span class="dt">selectionType=</span>fSelection2$selectionType)</code></pre></div>
</div>
<div id="longitudinal-data-analysis" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#longitudinal-data-analysis" class="anchor"> </a></body></html>Longitudinal Data Analysis</h1>
<p>Another feature of metavizr is to visualize data using a line plot. We detail the steps to perform this analysis and create a line plot using Metaviz such as those used in Paulson et al. when analyzing this time series data with a smoothing-spline [2] and [3].</p>
<p>First, import the etec16s dataset, select sample data from the first 9 days, and choose the feature annotations of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(etec16s)
<span class="kw">data</span>(etec16s)
etec16s &lt;-<span class="st"> </span>etec16s[,-<span class="kw">which</span>(<span class="kw">pData</span>(etec16s)$Day&gt;<span class="dv">9</span>)]
featuresOfInterest &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'Escherichia/Shigella'</span>,<span class="st">'Faecalibacterium prausnitzii'</span>)</code></pre></div>
<p>Next, use metagenomeSeq to fit a smoothing-spline to the time series data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">featureData</span>(etec16s)$Kingdom &lt;-<span class="st"> "Bacteria"</span>
feature_order &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"Kingdom"</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>, <span class="st">"Species"</span>, <span class="st">"OTU.ID"</span>)

timeSeriesFits &lt;-<span class="st"> </span><span class="kw">fitMultipleTimeSeries</span>(<span class="dt">obj=</span>etec16s,
                             <span class="dt">formula =</span> abundance~id +<span class="st"> </span>time*class +<span class="st"> </span>AntiGiven,
                             <span class="dt">class=</span><span class="st">"AnyDayDiarrhea"</span>,
                             <span class="dt">id=</span><span class="st">"SubjectID"</span>,
                             <span class="dt">time=</span><span class="st">"Day"</span>,
                             <span class="dt">lvl=</span><span class="st">"Species"</span>,
                             <span class="dt">feature_order =</span> feature_order,
                             <span class="dt">C=</span><span class="fl">0.3</span>,
                             <span class="dt">B=</span><span class="dv">1</span>)</code></pre></div>
<pre><code>## Warning in mspreg1(s, r, id.basis, y, wt, method, alpha, varht, random, :
## gss warning in ssanova: iteration for model selection fails to converge

## Warning in mspreg1(s, r, id.basis, y, wt, method, alpha, varht, random, :
## gss warning in ssanova: iteration for model selection fails to converge</code></pre>
<p>For plotting the data using Metaviz, we set the fit values as y-coordinates and timepoints as x-coordinates. We need to call <code>toMRexperiment</code> with arguments for the sample and feature data, in this case timepoints and annotations, respectively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">feature_order2 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"Kingdom"</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>, <span class="st">"Species"</span>)

splinesMRexp &lt;-<span class="st"> </span><span class="kw">ts2MRexperiment</span>(timeSeriesFits, <span class="dt">feature_data =</span> <span class="kw">featureData</span>(<span class="kw">aggregateByTaxonomy</span>(etec16s, <span class="dt">lvl=</span><span class="st">"Species"</span>, <span class="dt">feature_order =</span> feature_order2)), <span class="dt">sampleNames =</span> timeSeriesFits[[<span class="dv">2</span>]]$fit$timePoints)</code></pre></div>
<pre><code>##   [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
##  [18]  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34
##  [35]  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51
##  [52]  52  53  54  55   0  57  58  59  60  61  62  63  64  65  66  67  68
##  [69]  69  70  71  72  73  74  75   0  77  78  79  80  81  82  83  84  85
##  [86]  86  87  88  89  90  91  92  93  94  95   0   0  98  99 100 101 102
## [103] 103 104 105 106 107   0   0 110 111 112 113 114 115 116 117 118 119
## [120] 120 121 122 123 124   0 126 127 128   0 130   0 132 133 134 135 136
## [137] 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153
## [154] 154 155 156 157 158</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">etec16s_species &lt;-<span class="st"> </span><span class="kw">aggregateByTaxonomy</span>(etec16s, <span class="dt">lvl=</span><span class="st">"Species"</span>, <span class="dt">feature_order =</span> feature_order)</code></pre></div>
<p>Finally, we add the MRexperiment as a measurement for Metaviz to plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">splines_to_plot &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="dv">1</span>:<span class="kw">nrow</span>(<span class="kw">MRcounts</span>(splinesMRexp)), function(i) {<span class="kw">max</span>(<span class="kw">abs</span>(<span class="kw">MRcounts</span>(splinesMRexp[i,]))) &gt;=<span class="st"> </span><span class="dv">2</span>})
splines_to_plot_indices &lt;-<span class="st"> </span><span class="kw">which</span>(splines_to_plot ==<span class="st"> </span><span class="ot">TRUE</span>)

ic_plot &lt;-<span class="st"> </span>app$<span class="kw">plot</span>(splinesMRexp[splines_to_plot_indices,], <span class="dt">datasource_name =</span> <span class="st">"etec16_base"</span>, <span class="dt">control=</span><span class="kw"><a href="../reference/metavizControl.html">metavizControl</a></span>(<span class="dt">norm =</span> <span class="ot">FALSE</span>, <span class="dt">aggregateAtDepth =</span> <span class="dv">6</span>), <span class="dt">feature_order =</span> feature_order2)</code></pre></div>
<pre><code>## MRExperiment Object validated... PASS</code></pre>
<pre><code>## Warning in .Object$initialize(...): NaNs produced</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">splineObj &lt;-<span class="st"> </span>app$data_mgr$<span class="kw">add_measurements</span>(splinesMRexp[splines_to_plot_indices,], <span class="dt">datasource_name =</span> <span class="st">"splines"</span>, <span class="dt">control =</span> <span class="kw"><a href="../reference/metavizControl.html">metavizControl</a></span>(<span class="dt">norm=</span><span class="ot">FALSE</span>, <span class="dt">aggregateAtDepth =</span> <span class="dv">6</span>))</code></pre></div>
<pre><code>## MRExperiment Object validated... PASS</code></pre>
<pre><code>## Warning in .Object$initialize(...): NaNs produced</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">splineMeasurements &lt;-<span class="st"> </span>splineObj$<span class="kw">get_measurements</span>()
splineChart &lt;-<span class="st"> </span>app$chart_mgr$<span class="kw">visualize</span>(<span class="st">"LinePlot"</span>, splineMeasurements)</code></pre></div>
<p>We can update the colors and settings on the spline chart. For example, lets limit the y axis to be between -10 and 10. To do so we use the <code>set_chart_settings</code> method. We can list existing settings for a chart using the <code>list_chart_settings</code> function.</p>
<pre><code>##                           type                              js_class
## 1 epiviz.ui.charts.tree.Icicle          epiviz.ui.charts.tree.Icicle
## 2                  HeatmapPlot     epiviz.plugins.charts.HeatmapPlot
## 3                     LinePlot        epiviz.plugins.charts.LinePlot
## 4              StackedLinePlot epiviz.plugins.charts.StackedLinePlot
##   num_settings
## 1            5
## 2           17
## 3           16
## 4           14
##                                                                              settings
## 1                                 title,marginTop,marginBottom,marginLeft,marginRight
## 2 title,marginTop,marginBottom,marginLeft,marginRight,measurementGroupsAggregator,...
## 3 title,marginTop,marginBottom,marginLeft,marginRight,measurementGroupsAggregator,...
## 4 title,marginTop,marginBottom,marginLeft,marginRight,measurementGroupsAggregator,...
##   num_colors
## 1         10
## 2          8
## 3         20
## 4         20</code></pre>
<pre><code>##                             id                             label
## 1                        title                             Title
## 2                    marginTop                        Top margin
## 3                 marginBottom                     Bottom margin
## 4                   marginLeft                       Left margin
## 5                  marginRight                      Right margin
## 6  measurementGroupsAggregator Aggregator for measurement groups
## 7                     colLabel                    Columns labels
## 8                     rowLabel                        Row labels
## 9                   showPoints                       Show points
## 10                   showLines                        Show lines
## 11               showErrorBars                   Show error bars
## 12                 pointRadius                      Point radius
## 13               lineThickness                    Line thickness
## 14                        yMin                             Min Y
## 15                        yMax                             Max Y
## 16               interpolation                     Interpolation
##    default_value
## 1               
## 2             30
## 3             50
## 4             30
## 5             15
## 6     mean-stdev
## 7       colLabel
## 8           name
## 9          FALSE
## 10          TRUE
## 11          TRUE
## 12             4
## 13             3
## 14       default
## 15       default
## 16         basis
##                                                                                       possible_values
## 1                                                                                                    
## 2                                                                                                    
## 3                                                                                                    
## 4                                                                                                    
## 5                                                                                                    
## 6                                                              mean-stdev,quartiles,count,min,max,sum
## 7                                                                                                    
## 8                                                                                                    
## 9                                                                                                    
## 10                                                                                                   
## 11                                                                                                   
## 12                                                                                                   
## 13                                                                                                   
## 14                                                                                                   
## 15                                                                                                   
## 16 linear,step-before,step-after,basis,basis-open,basis-closed,bundle,cardinal,cardinal-open,monotone
##                      type
## 1                  string
## 2                  number
## 3                  number
## 4                  number
## 5                  number
## 6             categorical
## 7    measurementsMetadata
## 8  measurementsAnnotation
## 9                 boolean
## 10                boolean
## 11                boolean
## 12                 number
## 13                 number
## 14                 number
## 15                 number
## 16            categorical</code></pre>
<div class="figure">
<img src="images/Spline.png" alt="Spline metavizr"><p class="caption">Spline metavizr</p>
</div>
<div id="close-metavizr-and-end-session" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#close-metavizr-and-end-session" class="anchor"> </a></body></html>Close Metavizr and end session</h3>
<p>To close the connection with metaviz and the R session, we use the stop_app function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  app$<span class="kw">stop_app</span>()</code></pre></div>
</div>
<div id="sessioninfo" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#sessioninfo" class="anchor"> </a></body></html>SessionInfo</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.5 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] metavizr_0.99.5      etec16s_1.0.0        msd16s_0.106.0      
##  [4] metagenomeSeq_1.17.0 RColorBrewer_1.1-2   glmnet_2.0-5        
##  [7] foreach_1.4.3        Matrix_1.2-7.1       limma_3.28.21       
## [10] Biobase_2.32.0       BiocGenerics_0.18.0  devtools_1.12.0.9000
## 
## loaded via a namespace (and not attached):
##   [1] nlme_3.1-128               bitops_1.0-6              
##   [3] matrixStats_0.51.0         phyloseq_1.16.2           
##   [5] httr_1.2.1                 rprojroot_1.2             
##   [7] GenomeInfoDb_1.8.7         highlight_0.4.7           
##   [9] tools_3.3.1                backports_1.0.5           
##  [11] epivizr_2.5.1              vegan_2.4-1               
##  [13] R6_2.2.0                   KernSmooth_2.23-15        
##  [15] mgcv_1.8-16                colorspace_1.3-2          
##  [17] DBI_0.5-1                  lazyeval_0.2.0            
##  [19] permute_0.9-4              ade4_1.7-5                
##  [21] withr_1.0.2                curl_2.3                  
##  [23] git2r_0.18.0               graph_1.50.0              
##  [25] xml2_1.1.1                 desc_1.1.0                
##  [27] rtracklayer_1.32.2         scales_0.4.1              
##  [29] caTools_1.17.1             RBGL_1.48.1               
##  [31] pkgdown_0.1.0.9000         commonmark_1.1            
##  [33] stringr_1.1.0              digest_0.6.12             
##  [35] Rsamtools_1.24.0           epivizrStandalone_1.1.8   
##  [37] rmarkdown_1.3.9002         XVector_0.12.1            
##  [39] epivizrServer_1.3.1        htmltools_0.3.5           
##  [41] RSQLite_1.1-2              BiocInstaller_1.22.3      
##  [43] jsonlite_1.2               BiocParallel_1.6.6        
##  [45] gtools_3.5.0               RCurl_1.95-4.8            
##  [47] magrittr_1.5               biomformat_1.0.2          
##  [49] munsell_0.4.3              Rcpp_0.12.9               
##  [51] S4Vectors_0.10.3           ape_4.0                   
##  [53] stringi_1.1.2              whisker_0.3-2             
##  [55] yaml_2.1.14                MASS_7.3-45               
##  [57] SummarizedExperiment_1.2.3 zlibbioc_1.18.0           
##  [59] rhdf5_2.16.0               pkgbuild_0.0.0.9000       
##  [61] gplots_3.0.1               plyr_1.8.4                
##  [63] grid_3.3.1                 gdata_2.17.0              
##  [65] crayon_1.3.2               lattice_0.20-34           
##  [67] splines_3.3.1              Biostrings_2.40.2         
##  [69] multtest_2.28.0            GenomicFeatures_1.24.5    
##  [71] RNeo4j_1.6.4               knitr_1.15.1              
##  [73] igraph_1.0.1               GenomicRanges_1.24.3      
##  [75] rjson_0.2.15               reshape2_1.4.2            
##  [77] codetools_0.2-15           biomaRt_2.28.0            
##  [79] stats4_3.3.1               pkgload_0.0.0.9000        
##  [81] XML_3.98-1.5               evaluate_0.10             
##  [83] data.table_1.10.0          httpuv_1.3.3              
##  [85] testthat_1.0.2             gtable_0.2.0              
##  [87] purrr_0.2.2                assertthat_0.1            
##  [89] ggplot2_2.2.1              mime_0.5                  
##  [91] roxygen2_6.0.0             survival_2.40-1           
##  [93] tibble_1.2                 OrganismDbi_1.14.1        
##  [95] iterators_1.0.8            epivizrData_1.3.0         
##  [97] GenomicAlignments_1.8.4    AnnotationDbi_1.34.4      
##  [99] memoise_1.0.0              IRanges_2.6.1             
## [101] cluster_2.0.5              gss_2.1-6</code></pre>
<p>References:</p>
<p>[1] Pop, M., Walker, A.W., Paulson, J., Lindsay, B., Antonio, M., Hossain, M.A., Oundo, J., Tamboura, B., Mai, V., Astrovskaya, I. and Bravo, H.C., 2014. Diarrhea in young children from low-income countries leads to large-scale alterations in intestinal microbiota composition. Genome biology, 15(6), p.1.</p>
<p>[2] Pop, M., Paulson, J.N., Chakraborty, S., Astrovskaya, I., Lindsay, B.R., Li, S., Bravo, H.C., Harro, C., Parkhill, J., Walker, A.W. and Walker, R.I., 2016. Individual-specific changes in the human gut microbiota after challenge with enterotoxigenic Escherichia coli and subsequent ciprofloxacin treatment. BMC genomics, 17(1), p.1.</p>
<p>[3] Paulson J.N., Talukder H., and Bravo H.C, Longitudinal differential abundance analysis of microbial marker-gene surveys using smoothing splines. In Submission.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#preliminaries-the-data">Preliminaries: the data</a></li>
      <li><a href="#the-metavizr-session-manager">The metavizr session manager</a></li>
      <li><a href="#adding-an-icicle-navigation-widget-to-explore-metagenomic-features">Adding an Icicle navigation widget to explore metagenomic features</a></li>
      <li><a href="#visualizing-count-data-from-the-mousedata-mrexperiment-object">Visualizing count data from the <code>mouseData</code> MRExperiment object</a></li>
      <li><a href="#longitudinal-data-analysis">Longitudinal Data Analysis</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hector Corrada Bravo, Florin Chelaru, Justin Wagner, Jayaram Kancherla, Joseph Paulson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
